core/src/main/java/org/elasticsearch/search/profile/ProfileCollectorBuilder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.profile;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.MultiCollector;
+
+import java.util.AbstractList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Convenience object to help wrap collectors when profiling is enabled
+ */
+public class ProfileCollectorBuilder {
+
+    private final boolean profile;
+    private Collector currentCollector;
```

Comment:
this builder confuses me a lot. I think it should keep track of the current collector all the time not just if we are profiling.
I think we can have two static methods to acquire it like:

```Java

 public CollectorBuilder<InternalProfileCollector> getProfileCollectorBuilder() {
    return new CollectorBuilder<InternalProfileCollector>(true)
 }

 public CollectorBuilder<Collector> getCollectorBuilder() {
    return new CollectorBuilder<Collector>(false);
 }

public class ProfileCollectorBuilder<T extends Collector> {
  public void wrap(T collector, String reason);
  public void wrapMultiCollector(T collector, String reason, List<T> subCollectors);
}
```

that way the casting should not be needed.
we just keep on calling wrap / wrapMultiCollector until we are done and then call `public T build()` to get the collector?